File: mcspring-api/mcspring-test/src/main/java/in/kyle/mcspring/test/command/TestCommandRegistration.java
package in.kyle.mcspring.test.command;

import org.bukkit.command.CommandSender;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import in.kyle.mcspring.command.Command;
import in.kyle.mcspring.command.CommandRegistration;
import in.kyle.mcspring.command.CommandResolver;
import in.kyle.mcspring.command.SimpleMethodInjection;
import lombok.RequiredArgsConstructor;
import lombok.var;

@Component
@RequiredArgsConstructor
class TestCommandRegistration implements CommandRegistration {
    
    private final SimpleMethodInjection injection;
    private final Set<CommandResolver> commandResolvers;
    
    private final Map<String, Executor> commandExecutors = new HashMap<>();
    
    @Override
    public void register(Command command, Method method, Object object) {
        var executor = makeExecutor(method, object);
        getAllNames(command).forEach(key -> commandExecutors.put(key, executor));
    }
    
    private List<String> getAllNames(Command command) {
        List<String> commands = new ArrayList<>(Arrays.asList(command.aliases()));
        commands.add(command.value());
        return commands;
    }
    
    private Executor makeExecutor(Method method, Object object) {
        return (sender, label, args) -> {
            var temp = new CommandResolver.Command(sender, args, label);
            var contextResolvers = commandResolvers.stream()
                    .map(r -> r.makeResolver(temp))
                    .collect(Collectors.toList());
            Object result = injection.invoke(method, object, contextResolvers, sender, args, label);
            if (result != null) {
                sender.sendMessage(result.toString());
            }
        };
    }
    
    public void run(String command, CommandSender sender, String label, String[] args) {
        if (commandExecutors.containsKey(command)) {
            commandExecutors.get(command).execute(sender, label, args);
        } else {
            throw new RuntimeException(
                    "Command " + command + " is not registered. Make sure to @Import it");
        }
    }
    
    interface Executor {
        void execute(CommandSender sender, String label, String[] args);
    }
}

File: mcspring-api/mcspring-test/src/main/kotlin/in/kyle/mcspring/test/command/TestCommandRegistration.kt
package `in`.kyle.mcspring.test.command

import `in`.kyle.mcspring.command.Command
import `in`.kyle.mcspring.command.CommandRegistration
import `in`.kyle.mcspring.command.CommandResolver
import `in`.kyle.mcspring.command.SimpleMethodInjection
import org.bukkit.command.CommandSender
import org.springframework.stereotype.Component
import java.lang.reflect.Method
import java.util.*
import kotlin.reflect.KFunction
import kotlin.reflect.jvm.javaMethod
import kotlin.reflect.jvm.kotlinFunction

typealias CommandSig = (CommandSender, String, Array<String>) -> Unit

@Component
class TestCommandRegistration(
        private val injection: SimpleMethodInjection,
        private val commandResolvers: Set<CommandResolver>
) : CommandRegistration {

    private val commandExecutors: MutableMap<String, CommandSig> = HashMap()

    override fun register(command: Command, method: Method, obj: Any) {
        @Suppress("UNCHECKED_CAST")
        val executor = makeExecutor(method.kotlinFunction as KFunction<Any>, obj)
        command.aliases.plus(command.value).forEach { commandExecutors[it] = executor }
    }

    private fun makeExecutor(e: KFunction<Any>, obj: Any): CommandSig {
        return { sender: CommandSender, label: String, args: Array<String> ->
            val temp = CommandResolver.Command(sender, args.toList(), label)
            val miscResolvers = injection.makeResolvers(listOf(sender))
            val contextResolvers = commandResolvers.map { it.makeResolver(temp) }

            val parameterTypes = e.javaMethod!!.parameterTypes.toList()
            val parameters = injection.getParameters(parameterTypes, contextResolvers.plus(miscResolvers))
            val result = e.call(obj, *parameters)
            if (result !is Unit) {
                sender.sendMessage(result.toString())
            }
        }
    }

    fun run(command: String, sender: CommandSender, label: String, args: Array<String>) {
        if (command in commandExecutors) {
            commandExecutors[command]!!(sender, label, args)
        } else {
            throw RuntimeException("Command $command is not registered. Make sure to @Import it")
        }
    }
}
